routing/pathfind.go

```diff
@@ -534,82 +550,99 @@ func findPath(tx *bolt.Tx, graph *channeldb.ChannelGraph,
 		// exploring this edge.
-		if _, ok := ignoredNodes[v]; ok {
+		if _, ok := ignoredNodes[fromVertex]; ok {
 			return
 		}
 		if _, ok := ignoredEdges[edge.ChannelID]; ok {
 			return
 		}
 
+		nodeInfo := distance[toNode]
+
+		amountToSend := nodeInfo.amountToReceive
+
+		// Compute fee based on the amount that needs to be send to the
+		// next node.
+		var fee lnwire.MilliSatoshi
+		if fromVertex != sourceVertex {
+			fee = computeFee(amountToSend, edge)
+		} else {
+			// Source node has no precedessor to pay fee
+			// and fee should therefore also not be included in
+			// fee limit check and edge weight.
+			fee = 0
+		}
+
+		// Check if accumulated fees would exceed fee limit when
+		// this node would be added to the path.
+		totalFee := amountToSend + fee - amt
+		if totalFee > feeLimit {
+			return
+		}
+
 		// Compute the tentative distance to this new channel/edge which
-		// is the distance to our pivot node plus the weight of this
-		// edge.
-		tempDist := distance[pivot].dist + edgeWeight(amt, edge)
+		// is the distance from our toNode to the target node plus the 
+		// weight of this edge.
+		tempDist := nodeInfo.dist + edgeWeight(fee, edge.TimeLockDelta)
 
 		// If this new tentative distance is better than the current
 		// best known distance to this node, then we record the new
 		// better distance, and also populate our "next hop" map with
 		// this edge. We'll also shave off irrelevant edges by adding
 		// the sufficient capacity of an edge and clearing their
 		// min-htlc amount to our relaxation condition.
-		if tempDist < distance[v].dist && bandwidth >= amt &&
-			amt >= edge.MinHTLC && edge.TimeLockDelta != 0 {
+		if tempDist < distance[fromVertex].dist && bandwidth >= amountToSend &&
+			amountToSend >= edge.MinHTLC && edge.TimeLockDelta != 0 {
+
+			amountToReceive := amountToSend + fee
 
-			distance[v] = nodeWithDist{
-				dist: tempDist,
-				node: edge.Node,
+			distance[fromVertex] = nodeWithDist{
+				dist:            tempDist,
+				node:            fromNode,
+				amountToReceive: amountToReceive,
+				fee:             fee,
 			}
 
-			prev[v] = edgeWithPrev{
-				edge: &ChannelHop{
-					ChannelEdgePolicy: edge,
-					Capacity:          bandwidth.ToSatoshis(),
-				},
-				prevNode: pivot,
+			next[fromVertex] = &ChannelHop{
+				ChannelEdgePolicy: edge,
+				Bandwidth:         bandwidth,
 			}
 
 			// Add this new node to our heap as we'd like to further
-			// explore down this edge.
-			heap.Push(&nodeHeap, distance[v])
+			// explore backwards through this edge.
+			heap.Push(&nodeHeap, distance[fromVertex])
 		}
 	}
 
 	// TODO(roasbeef): also add path caching
 	//  * similar to route caching, but doesn't factor in the amount
 
-	// To start, we add the source of our path finding attempt to the
-	// distance map with a distance of 0. This indicates our starting
-	// point in the graph traversal.
-	sourceVertex := Vertex(sourceNode.PubKeyBytes)
-	distance[sourceVertex] = nodeWithDist{
```

Comment:
we must still make sure the source is added to the map with dist `infinity`

Reply:
The source node is already added in the graph.ForEachNode loop further up. The reason it was explictly added here previously, was to set (overwrite) the distance to 0. But because of backwards searching, this is no longer necessary.

Reply:
Hmm, we are then assuming that the source node will always be returned from `graph.ForEachNode`. That's a reasonable assumption to make, but I think it won't hurt to also handle the case where it's not.

Maybe that's fine anyway, because we will just finish or path finding, and then discover no path was found to the source node? Or do we depend on it being in the map?

Reply:
It is always returned. If you look at 

https://github.com/lightningnetwork/lnd/blob/b5a228808b0446d6758a23e24d6c686aab9936fc/channeldb/graph.go#L276

it can be seen that even channeldb itself relies on the source pubkey being present in the node bucket.

If these semantics would change at some point, path finding would never find a path anymore. So the bug that would be created, will quickly surface.

I'd prefer not to add a check anyway, because it can only elicit questions from developer about the reason there might be for this to happen. 

Comment added.